drivers/serial/uart_native_pty.c

```diff
@@ -0,0 +1,320 @@
+#define DT_DRV_COMPAT zephyr_native_posix_uart
+#warning "zephyr,native-posix-uart is deprecated in favor of zephyr,native-pty-uart"
+#else
+#define DT_DRV_COMPAT zephyr_native_pty_uart
+#endif
+
+#include <stdbool.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/kernel.h>
+#include <cmdline.h> /* native_sim command line options header */
+#include <posix_native_task.h>
+#include <nsi_host_trampolines.h>
+#include <nsi_tracing.h>
+#include "uart_native_pty_bottom.h"
+
+#define ERROR posix_print_error_and_exit
+#define WARN posix_print_warning
+
+/*
+ * UART driver for native simulator based boards.
+ * It can support a configurable number of UARTs.
+ *
+ * One (and only one) of this can be connected to the process STDIN+STDOUT otherwise, they are
+ * connected to a dedicated pseudo terminal.
+ *
+ * Connecting to a dedicated PTY is the recommended option for interactive use, as the pseudo
+ * terminal driver will be configured in "raw" mode and will therefore behave more like a real UART.
+ *
+ * When connected to its own pseudo terminal, it may also auto attach a terminal emulator to it,
+ * if set so from command line.
+ */
+
+struct native_pty_status {
+	int out_fd;       /* File descriptor used for output */
+	int in_fd;        /* File descriptor used for input */
+	bool on_stdinout; /* This UART is connected to a PTY and not STDIN/OUT */
+
+	bool auto_attach;      /* For PTY, attach a terminal emulator automatically */
+	char *auto_attach_cmd; /* If auto_attach, which command to launch the terminal emulator */
+	bool wait_pts;         /* Hold writes to the uart/pts until a client is connected/ready */
+	bool cmd_request_stdinout; /* User requested to connect this UART to the stdin/out */
+};
+
+static void np_uart_poll_out(const struct device *dev, unsigned char out_char);
+static int np_uart_poll_in(const struct device *dev, unsigned char *p_char);
+static int np_uart_init(const struct device *dev);
+
+static DEVICE_API(uart, np_uart_driver_api) = {
+	.poll_out = np_uart_poll_out,
+	.poll_in = np_uart_poll_in,
+};
+
+#define NATIVE_PTY_INSTANCE(inst)                                        \
+	static struct native_pty_status native_pty_status_##inst;        \
+								         \
+	DEVICE_DT_INST_DEFINE(inst, np_uart_init, NULL,                  \
+			      (void *)&native_pty_status_##inst, NULL,   \
+			      PRE_KERNEL_1, CONFIG_SERIAL_INIT_PRIORITY, \
+			      &np_uart_driver_api);
+
+DT_INST_FOREACH_STATUS_OKAY(NATIVE_PTY_INSTANCE);
+
+/**
+ * @brief Initialize a native_pty serial port
+ *
+ * @param dev uart device struct
+ *
+ * @return 0 (if it fails catastrophically, the execution is terminated)
+ */
+static int np_uart_init(const struct device *dev)
+{
+	static bool first_node = true;
+	static bool stdinout_used;
+	struct native_pty_status *d;
+
+	d = (struct native_pty_status *)dev->data;
+
+	if (first_node && IS_ENABLED(CONFIG_UART_NATIVE_PTY_0_ON_STDINOUT)) {
+		d->on_stdinout = true;
+	}
+	first_node = false;
```

Comment:
It seems that `first_node` is only used if `CONFIG_UART_NATIVE_PTY_0_ON_STDINOUT` is set.
So we could do here this to avoid defining `first_node` in the first place.
```
if (IS_ENABLED(CONFIG_UART_NATIVE_PTY_0_ON_STDINOUT)) {
	static bool first_node = true;

	if (first_node) {
		d->on_stdinout = true;
		first_node = false;
	}
}
```

Reply:
Done.